resource limiting: complete rework to handle context cancelation and native rate-limiting on a template (#196)
ovh/utask: README.md

foreach: '{{.step.prefixStrings.children | toJson}}'
```

#### Resources <a name="resources"></a>

> Resources is a way to limit the number of current execution of tasks or steps within an µTask instance. When a lot of tasks are launched at the same time, a high throughput from µTask can hurt some components of your infrastructure, making them unavailable.

Reviewer: ```suggestion
Resources is a way to limit the number of concurrent execution of tasks or steps within an µTask instance. When a lot of tasks are launched at the same time, a high throughput from µTask can hurt some components of your infrastructure, making them unavailable.
```

> Resources is a way to limit the number of current execution of tasks or steps within an µTask instance. When a lot of tasks are launched at the same time, a high throughput from µTask can hurt some components of your infrastructure, making them unavailable.

Reviewer 2: Something like.

Resources are a way to restrict the concurrency factor of certain operations, to control the throughput and avoid dangerous behavior e.g. flooding the targets.

High level view:
- For each action to execute, a list of target `resources` is determined. (see later)
- In the uTask configuration, limits can be set for each `resource` label. This acts as a semaphore, allowing a certain number of concurrent slots for the given `resource` label. If no limit is set for a `resource` label, the previously mentionned target `resources` have no effect.

The target `resources` for a step can be defined in its YAML definition, using the `resources` property.
Alternatively, some target `resources` are determined automatically:
- When a task is run, the `resource` `template:my-template-name` is used automatically
- When a step is run, the plugin in charge of the execution automatically generates a list of `resources`. This includes generic `resources` such as `socket`, `url:www.google.com`, `fork`, ... allowing the uTask administrator to set-up generic limits such as `"socket": 48` or `"url:www.google.com": 1`.

> Resources is a way to limit the number of current execution of tasks or steps within an µTask instance. When a lot of tasks are launched at the same time, a high throughput from µTask can hurt some components of your infrastructure, making them unavailable.

Author: done

> Defining `resources` on a µTask instance, on a template, or on a step is a way to limit the number of concurrent executions on µTask side, to make sure that the targetted component stays healthy.

Reviewer: ```suggestion
Defining `resources` on a µTask instance, on a template, or a step is a way to limit the number of concurrent executions on µTask side, to make sure that the targetted component stays healthy.
```

µTask comes with some default resources: builtin plugins declares some resources, which can be throttled at usage, in the µTask configuration. For example, the `http` plugin declares two resources: `socket` and `url:example.org` (with `example.org` being the hostname of the destination). If you want to limit the maximum outgoing open socket on a µTask instance, you can add a `resource_limits` for `socket` in the configuration, the value being the number of concurrent usage allowed per instance.

It's also possible to declare a `resource` on a step, inside the template.

```yaml
steps:
  foobar:
    description: A dummy step, that should not execute in parallel
    resources: [limitedResourceFoobar]
    action:
      type: echo
      configuration:
        output:
          foobar: fuzz
```

> Every time the step `foobar` will be executed, µTask engine will verify that the maximum concurrent limit is not exceeded, and acquire one slot. The slot will be hold until the step finishes its execution. If all slots for a resource are hold, then the current execution will wait until one slot is released, or if timeout is reached (defined in `resource_acquire_timeout` configuration, or default `1 minute`). If the resource is not declared in the `resource_limits` configuration, then, no limitation will be applied. If the limitation value is set to `0`, then, the step will not wait, and will be set to state `TO_RETRY` to be schedule later.

Reviewer: ```suggestion
Every time the step `foobar` will be executed, µTask engine will verify that the maximum concurrent limit is not exceeded, and acquire one slot. This slot will be held until the step finishes its execution. If all slots for a resource are held, then the current execution will wait until one slot is released, or if the timeout is reached (defined in `resource_acquire_timeout` configuration, or default `1 minute`). If the resource is not declared in the `resource_limits` configuration, then, no limitation will be applied. If the limit value is set to `0`, then, the step will not wait and will be set to state `TO_RETRY` to be scheduled later.
```

> Templates declare a builtin resource, named `template:my-template-name` (with `my-template-name` being the template name): this allows to limit the number of concurrent execution of tasks from the given template. In that way, a way that consumes a lot of CPU/RAM/IO can be restrain to save the system resources, or to disable completely the execution of tasks from a given template.

Reviewer: ```suggestion
Templates declare a builtin resource, named `template:my-template-name` (with `my-template-name` being the template name): this allows to limit the number of concurrent execution of tasks from the given template. In that way, a way that consumes a lot of CPU/RAM/IO can be restrained to save the system resources or to disable completely the execution of tasks from a given template.
```

foreach: '{{.step.prefixStrings.children | toJson}}'
```

> #### Resources <a name="resources"></a>

Reviewer 3: What is the link used for ?

Author: It's used line 359 to give a quick link to this section.
